statsmodels/base/_penalties.py

```diff
@@ -110,13 +110,190 @@ def func(self, params):
         v *= self.dlt**2
-        return np.sum(self.wts * self.alpha * v)
+        return np.sum(self.wts * self.alpha * v, 0)
 
     def grad(self, params):
         v = np.sqrt(1 + (params / self.dlt)**2)
         return params * self.wts * self.alpha / v
 
 
+class SCAD(Penalty):
+    """
+    The SCAD penalty of Fan and Li
+    """
+
+    def __init__(self, tau, c=3.7, wts=None):
+        if wts is None:
+            self.weights = 1.
+        else:
+            self.weights = wts
+        self.tau = tau
+        self.c = c
+
+    def func(self, params):
+
+        # 3 segments in absolute value
+        tau = self.tau
+        p_abs = np.atleast_1d(np.abs(params))
+        res = np.empty(p_abs.shape)#, p_abs.dtype)
+        res.fill(np.nan)
+        mask1 = p_abs < tau
+        mask3 = p_abs >= self.c * tau
+        res[mask1] = tau * p_abs[mask1]
+        mask2 = ~mask1 & ~mask3
+        p_abs2 = p_abs[mask2]
+        #tmp = (tau * p_abs2**2 - 2 * self.c * p_abs2 + tau**2)
+        tmp = (p_abs2**2 - 2 * self.c * tau * p_abs2 + tau**2)
+        res[mask2] = -tmp / (2 * (self.c - 1))
+        #res[mask3] = (self.c + 1) / 2. * p_abs[mask3]**2
+        res[mask3] = (self.c + 1) * tau**2 / 2.
+
+        return (self.weights * res).sum(0)
+
+    def grad(self, params):
+
+        # 3 segments in absolute value
+        tau = self.tau
+        p = np.atleast_1d(params)
+        p_abs = np.abs(p)
+        p_sign = np.sign(p)
+        res = np.empty(p_abs.shape)#, p_abs.dtype)
+        res.fill(np.nan)
+
+        mask1 = p_abs < tau
+        mask3 = p_abs >= self.c * tau
+        mask2 = ~mask1 & ~mask3
+        res[mask1] = p_sign[mask1] * tau
+        #tmp = p_sign[mask2] * (tau * p_abs[mask2] - self.c)
+        tmp = p_sign[mask2] * (p_abs[mask2] - self.c * tau)
+        res[mask2] = -tmp / (self.c - 1)
+        res[mask3] = 0#(self.c + 1)
+
+        return self.weights * res
+
+
+    def deriv2(self, params):
+        """Second derivative of function
+
+        Warning: Not checked, possible problems in multivariate case
+        This returns scalar or vector in same shape as params, not a square Hessian.
+
+        """
+
+        # 3 segments in absolute value
+        tau = self.tau
+        p = np.atleast_1d(params)
+        p_abs = np.abs(p)
+        p_sign = np.sign(p)
+        res = np.zeros(p_abs.shape)#, p_abs.dtype)
+
+        mask1 = p_abs < tau
+        mask3 = p_abs >= self.c * tau
+        mask2 = ~mask1 & ~mask3
+
+        res[mask2] = -p_sign[mask2] / (self.c - 1)
+
+        return self.weights * res
+
+
+class SCADSmoothed(SCAD):
+    """
+    The SCAD penalty of Fan and Li, quadratically smoothed around zero
+
+    This follows Fan and Li 2001 equation (3.7).
+
+    Parameterization follows Boo, Johnson, Li and Tan 2011
+
+
+    TODO: Use delegation instead of subclassing, so smoothing can be added to all
+    penalty classes.
+
+    """
+
+    def __init__(self, tau, c=3.7, c0=None, wts=None, restriction=None):
```

Comment:
`wts` or `weights`?

Reply:
internally I use weights because it looks better

Reply:
Should have been changed to weights, will be now.